voxel_grid/prototype/mvp_v2.hpp

```diff
@@ -0,0 +1,254 @@
+
+
+#include <Eigen/Dense>
+#include <boost/optional.hpp>
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+
+using Vec3f = Eigen::Vector3f;
+
+class PointXYZ {
+ public:
+  PointXYZ(const float x = 0, const float y = 0, const float z = 0) {
+    p << x, y, z;
+  }
+  PointXYZ(const Vec3f &p) : p(p) {}
+  Vec3f p;
+};
+
+bool isXYZFinite(const PointXYZ &pt) { return true; }
+
+template <typename PointT>
+class PointCloudT {
+ public:
+  PointCloudT() {}
+
+  void push_back(const PointT &p) { points.push_back(p); }
+  void reserve(const size_t s) { (void)s; }
+  auto begin() { return points.begin(); }
+  auto end() { return points.end(); }
+
+  // Defined in pcl::PointCloud
+  using PointType = PointT;
+
+  bool is_dense;
+  std::vector<PointT> points;
+};
+
+template <typename PointT>
+class Filter {
+  using PointCloud = PointCloudT<PointXYZ>;
+
+ public:
+  Filter() {}
+  void filter(PointCloud &output) { applyFilter(output); }
+
+ protected:
+  virtual void applyFilter(PointCloud &pc) = 0;
+  PointCloud input_;
+};
+
+// ================== transform_filter.hpp ==================
+
+#define GET_POINT_TYPE(GridStructT) typename GridStructT::PointCloud::PointType
+
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class TransformFilter : public Filter<PointT> {
+  using PointCloud = PointCloudT<PointXYZ>;
+
+ protected:
+  TransformFilter(GridStruct grid_struct) {
+    grid_struct_ = std::move(grid_struct);
+  }
+
+  void applyFilter(PointCloud &output) {
+    if (!grid_struct_.setUp(this)) {
+      output = this->input_;
+      return;
+    }
+
+    for (const auto &pt : this->input_) {
+      if (this->input_.is_dense || isXYZFinite(pt)) {
+        grid_struct_.addPointToGrid(pt);
+      }
+    }
+
+    // allocate enough space for points
+    output.reserve(grid_struct_.size());
+    for (auto it = grid_struct_.begin(); it != grid_struct_.end(); ++it) {
+      const auto &res = grid_struct_.filterGrid(it);
+      if (res) {
+        output.push_back(res.value());
+      }
+    }
+
+    // push back leaf_layout_ to the base
+    grid_struct_.saveResults(this);
```

Comment:
We can prevent the static_cast and this function requirement for all filters by moving this into `void GridFilterBase::applyFilter(PointCloud& out) override`

Reply:
I can try it out. 

Then the purposes of `TransformFilter` are just holding `GridStruct` object and accessing the data in `GridStruct` right?

Reply:
I removed `saveResults` so I'll keep `applyFilter` in TransformFilter` for now (`static_cast` is still needed)